### PR TITLE
Reduction kernel optimization

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Optimize reduction kernel code by moving loads from global memory before computation.
Add CMake option to build CUDA code with --generate-line-info option.

Results from running float reduce_square_sum() 1000 times:

number of elements | baseline average ms | updated average ms | baseline/updated
-- | -- | -- | --
1024 | 0.005959 | 0.005421 | 1.09931
2048 | 0.008346 | 0.007856 | 1.062425
4096 | 0.008491 | 0.007929 | 1.070858
8192 | 0.008609 | 0.008151 | 1.056164
1048576 | 0.012841 | 0.011321 | 1.134248
1073741824 | 8.312 | 5.12766 | 1.621012

**Motivation and Context**
Address performance regression observed in training throughput tests from #5639.

### CODE REVIEW

**Comment by weixingzhang** [comment]
Have the training throughput tests been merged in ORT repo?

**Comment by edgchen1** [comment]
> Have the training throughput tests been merged in ORT repo? If not, I am wondering whether these tests can be merged in ORT repo to catch the regression for the future change.

Yes, these are the tests that show up on the performance dashboard.
https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/azure-pipelines/orttraining-linux-gpu-perf-test-ci-pipeline.yml

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cu [hunk 1]
```diff
@@ -134,26 +134,37 @@ __device__ void reduce_all(
...
+    } else {
+      output[0] = TFinalOp()(result);
+    }
+  };
+
   // Thread-level reduction (storage change: global memory -> register).
   // One thread reduces MAX_NUM_ELEMENTS_PER_THREAD elements to a thread register
   // in one iteration.
   TBuf value = 0;
   for (int id = tid_in_grid_row; id < num_elements; id += MAX_NUM_ELEMENTS_PER_THREAD * num_threads_in_grid_row) {
```
Comment (wschin) [inline onnxruntime/core/providers/cuda/reduction/reduction_functions.cu:151]: Move this declaration to Ln 161.

**Review by wschin — APPROVED** [review]
:shipit:

File: onnxruntime/core/providers/cuda/reduction/reduction_functions.cu (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/providers/cuda/reduction/reduction_functions.cu:151]: this is used in line 158, will leave it here

**Pull request merged**